### ISSUE TITLE
Order of lemmas returned by derivationally_related_forms() of WordNet?

### ISSUE THREAD

**Issue by ymfa** [nltk/nltk#1805]
I use `derivationally_related_forms()` to get the derivationally related lemmas of a WordNet lemma. When I upgraded nltk from version 2.0.4 to 3.2.3, I realized that the lemmas returned by this function are no longer sorted as they were. Instead, the order could change from one Python session to another. Of course, I could sort the list afterwards to get a deterministic result, but I want to replicate the same order as I found in nltk 2 (which doesn't work on Python 3).

The following code is an example. Please save it as a script, and run it a few times to see if the output stays the same for each run. (Remember that for nltk 2, you need to change `synset.lemmas()` to `synset.lemmas` on the third line.)

    from nltk.corpus import wordnet as wn
    synset = wn.synsets('study', 'n')[1]
    lemma = synset.lemmas()[0]
    print(lemma.derivationally_related_forms())

The output from nltk 2 is always:
`[Lemma('study.v.02.study'), Lemma('study.v.05.study'), Lemma('bookish.s.01.studious'), Lemma('learn.v.04.study')]`

However, according to my observation, the output from nltk 3 is either:
`[Lemma('bookish.s.01.studious'), Lemma('study.v.05.study'), Lemma('study.v.02.study'), Lemma('learn.v.04.study')]`
Or:
`[Lemma('bookish.s.01.studious'), Lemma('study.v.02.study'), Lemma('learn.v.04.study'), Lemma('study.v.05.study')]`

Does anyone know how was the list ordered in nltk 2? Does the order have any linguistic significance? A temporary workaround that allows me to replicate the same order in nltk 3 would be appreciated.

**Comment by alvations** [comment]
@ymfa Which version of Python3 and Python2 are you using? 

    python3 -c "import sys; print(sys.version)"
    python2 -c "import sys; print(sys.version)"

Try upgrading your Python version to 3.4 or 3.5.

Currently, NLTK supports Python2.7 and Python 3.4 and Python3.5. 
Python 3.6 should work (most of the time) too but regression tests might fail. 

On my Ubuntu machine, they return the same list (Python 2.7.12 and Python 3.5.2):

```
$ python2

Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from nltk.corpus import wordnet as wn

>>> study = wn.synsets('study', 'n')[1].lemmas()[0]
>>> study.derivationally_related_forms()
[Lemma('bookish.s.01.studious'), Lemma('study.v.02.study'), Lemma('study.v.05.study'), Lemma('learn.v.04.study')]
>>> exit()

$ python3
Python 3.5.2 (default, Nov 17 2016, 17:05:23) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from nltk.corpus import wordnet as wn
>>> study = wn.synsets('study', 'n')[1].lemmas()[0]
>>> study.derivationally_related_forms()
[Lemma('bookish.s.01.studious'), Lemma('study.v.05.study'), Lemma('study.v.02.study'), Lemma('learn.v.04.study')]
>>> exit()
```

On Mac (Python 2.7.13 and Python 3.6.1):

```
$ python
Python 2.7.13 (default, Dec 18 2016, 07:03:39) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from nltk.corpus import wordnet as wn
>>> study = wn.synsets('study', 'n')[1].lemmas()[0]
>>> study.derivationally_related_forms()
[Lemma('bookish.s.01.studious'), Lemma('study.v.02.study'), Lemma('study.v.05.study'), Lemma('learn.v.04.study')]
>>> exit()

$ python3
Python 3.6.1 (default, Apr  4 2017, 09:40:21) 
[GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from nltk.corpus import wordnet as wn
>>> study = wn.synsets('study', 'n')[1].lemmas()[0]
>>> study.derivationally_related_forms()
[Lemma('bookish.s.01.studious'), Lemma('study.v.02.study'), Lemma('learn.v.04.study'), Lemma('study.v.05.study')]
>>> exit()
```

**Comment by ymfa** [comment]
@alvations These are the Python versions I use.
```
$ python3 -c "import sys; print(sys.version)"
3.6.1 |Anaconda 4.4.0 (x86_64)| (default, May 11 2017, 13:04:09) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)]

$ /usr/bin/python -c "import sys; print(sys.version)"
2.7.10 (default, Feb  7 2017, 00:08:15) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]
```

Please note that in tests you just posted (both on Ubuntu and on Mac), the lists are not in the same order.

When I looked at the source code, I realized that in NLTK 3 there is a `sorted()` wrapped around the return value of `Lemma._related()` in nltk/nltk/corpus/reader/wordnet.py, which didn't exist in NLTK 2:
```
    def _related(self, relation_symbol):
        get_synset = self._wordnet_corpus_reader.synset_from_pos_and_offset
        return sorted([
            get_synset(pos, offset)._lemmas[lemma_index]
            for pos, offset, lemma_index
            in self._synset._lemma_pointers[self._name, relation_symbol]
        ])
```
However, removing the `sorted()` does not revert it to NLTK 2 behaviour either, and it still returns lemmas in different order.

**Comment by alvations** [comment]
@ymfa You're right the "list" of lemmas were not ordered.

The problem looks a little spaghetti-like:

The relations comes from the [`_related()`](https://github.com/nltk/nltk/blob/develop/nltk/corpus/reader/wordnet.py#L299) function which seem to return a `sorted()` list but the thing that its sorting neither the lemma name string nor the synset index from `get_synset()` function.

```python
    def _related(self, relation_symbol):
        get_synset = self._wordnet_corpus_reader.synset_from_pos_and_offset
        return sorted([
            get_synset(pos, offset)._lemmas[lemma_index]
            for pos, offset, lemma_index
            in self._synset._lemma_pointers[self._name, relation_symbol]
        ])
````

Looking at the [`synset_from_pos_and_offset()`](https://github.com/nltk/nltk/blob/develop/nltk/corpus/reader/wordnet.py#L1327), we see that the `sorted([get_synset(...)])` from `_related()` function is sorting the hash within the nested dictionary hash from `self._synset_offset_cache[pos][offset]`:

```python
    def synset_from_pos_and_offset(self, pos, offset):
        # Check to see if the synset is in the cache
        if offset in self._synset_offset_cache[pos]:
            return self._synset_offset_cache[pos][offset]

        data_file = self._data_file(pos)
        data_file.seek(offset)
        data_file_line = data_file.readline()
        synset = self._synset_from_pos_and_line(pos, data_file_line)
        assert synset._offset == offset
        self._synset_offset_cache[pos][offset] = synset
        return synset
```


Before the ["death of OrderedDict" patch](http://bugs.python.org/issue27350), the order of a sorted dictionary isn't promised.  

Other than waiting for Python <3.5 to phase out naturally and #1659 to be patched. 
Any suggestions as to how we can patch this on our end?

**Comment by alvations** [comment]
This bug would be phase out by the new `dict` object in Python 3.6. Dictionary will be always ordered Python 3.6 by @methane and @rhettinger

- https://mail.python.org/pipermail/python-dev/2016-September/thread.html#146327
- https://twitter.com/raymondh/status/850102884972675072

**Comment by alvations** [comment]
After support for Python 3.6, moving forward, I think the issue is resolved upstream in CPython and also in #1918. 

Thanks @ymfa !
